// windmill/engine/server/posts.js
var fs = require('fs'),
    path = require('path'),
    markdown = require('./markdown');

var posts_directory = path.join(__dirname + '/../../posts');

module.exports = {
    getAll: function(done) {
        var posts = [];
        var files = fs.readdirSync(posts_directory);
    
        files.forEach(function(file){
            posts.push(preparePost(file));
        });
        
        return posts;
    },
    
    getByName: function(file) {
        return preparePost(file);
    }
};

function preparePost(file) {
    var content = fs.readFileSync(path.join(posts_directory, file), 'utf-8');
    
    // 5 first lines = post info
    var lines = content.split("\n");
    var header = '';
    for (var i = 0; i <= 5; i++) {
        header = header + lines[i];
    }
    
    var post_content = '';
    for (var i = 5; i < lines.length; i++) {
        post_content = post_content + "\n" + lines[i];
    }
    
    // parsing 5 first lines
    header = JSON.parse(header);
    
    // posts infos
    var title = header.title;
    var author = header.author;
    var published = header.published;
    
    // todo: use filename as date
    
    // post object
    var post = {
        title: title,
        author: author,
        published: published.toString(),
        url: file.replace('.post', ''),
        content: markdown.parse(post_content) 
    };
    
    return post;
}
